src/extension.jsligo

```diff
@@ -0,0 +1,138 @@
+#import "ligo-generic-fa2/lib/multi_asset/fa2.mligo" "FA2"
+#import "./errors.jsligo" "Errors"
+
+type seconds = nat;
+export type permitKey = [address, bytes];
+export type permits = big_map<permitKey, timestamp>;
+export type userExpiries = big_map<address, option<seconds>>;
+export type permitExpiries = big_map<permitKey, option<seconds>>;
+
+// token_id -> total_supply
+export type tokenTotalSupply = big_map<FA2.Ledger.token_id, nat>;
+
+export type t = {
+    admin: address,
+    counter: nat,
+    defaultExpiry: seconds,
+    maxExpiry: seconds,
+    permits: permits,
+    userExpiries: userExpiries,
+    permitExpiries: permitExpiries,
+    tokenTotalSupply : tokenTotalSupply
+};
+
+let get_user_defined_expiry = ([from_, ext]: [address, t]) : option<seconds> => {
+    return match (Big_map.find_opt(from_, ext.userExpiries), {
+        None: () => (Some(ext.defaultExpiry)),
+        Some: (exp: option<seconds>) => exp
+    });
+};
+
+let getExpiry = ([ext, paramPermitKey]: [t, permitKey]) : seconds => {
+    match ((Big_map.find_opt([paramPermitKey[0], paramPermitKey[1]], ext.permitExpiries) as option<option<seconds>>), {
+        None: () => {
+                match ((Big_map.find_opt(paramPermitKey[0], ext.userExpiries) as option<option<seconds>>), {
+                    None: () => ext.defaultExpiry,
+                    Some: (exp: option<seconds>) => {
+                        match (exp, {
+                            None: () => ext.defaultExpiry,
+                            Some: (t: seconds) => t
+                        })
+                    }
+                })
+            },
+        Some: (p: option<seconds>) => {
+                match (p, {
+                    None: () => ext.defaultExpiry,
+                    Some: (exp: seconds) => exp
+                })
+            }
+    })
+};
+
+
+export let assertAdmin = (ext : t) : unit =>
+    assert_with_error((Tezos.get_sender() == ext.admin), Errors.requiresAdmin);
+
+export let setAdmin = ([ext, admin] : [t, address]) : t => {
+    assertAdmin(ext);
+    return { ...ext, admin : admin };
+};
+
+let getSupply = ([supply, tokenId] : [tokenTotalSupply, FA2.Ledger.token_id]) : nat =>
```

Comment:
Why have we such a strong link between the extension implementation and the FA2?

Reply:
agreed Supply managment has been moved to a separate file